service/src/escrow_monitor.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use std::sync::Arc;
+
+use alloy_primitives::Address;
+use anyhow::Result;
+use ethereum_types::U256;
+use log::{error, info};
+
+use serde::Deserialize;
+
+use tokio::sync::RwLock;
+
+use crate::common::types::GraphQLQuery;
+use crate::graph_node::GraphNodeInstance;
+
+#[derive(Debug)]
+struct EscrowMonitorInner {
+    graph_node: GraphNodeInstance,
+    escrow_subgraph_deployment: String,
+    indexer_address: Address,
+    interval_ms: u64,
+    sender_accounts: Arc<RwLock<HashMap<Address, U256>>>,
+}
+
+#[cfg_attr(test, faux::create)]
+#[derive(Debug, Clone)]
+pub struct EscrowMonitor {
+    _monitor_handle: Arc<tokio::task::JoinHandle<()>>,
+    inner: Arc<EscrowMonitorInner>,
+}
+
+#[cfg_attr(test, faux::methods)]
+impl EscrowMonitor {
+    pub async fn new(
+        graph_node: GraphNodeInstance,
+        escrow_subgraph_deployment: String,
+        indexer_address: Address,
+        interval_ms: u64,
+    ) -> Result<Self> {
+        let sender_accounts = Arc::new(RwLock::new(HashMap::new()));
+
+        let inner = Arc::new(EscrowMonitorInner {
+            graph_node,
+            escrow_subgraph_deployment,
+            indexer_address,
+            interval_ms,
+            sender_accounts,
+        });
+
+        let inner_clone = inner.clone();
+
+        let monitor = EscrowMonitor {
+            _monitor_handle: Arc::new(tokio::spawn(async move {
+                EscrowMonitor::monitor_loop(&inner_clone).await.unwrap();
+            })),
+            inner,
+        };
+
+        Ok(monitor)
+    }
+
+    async fn current_accounts(
+        graph_node: &GraphNodeInstance,
+        escrow_subgraph_deployment: &str,
+        indexer_address: &Address,
+    ) -> Result<HashMap<Address, U256>> {
+        // These 2 structs are used to deserialize the response from the escrow subgraph.
+        // Note that U256's serde implementation is based on serializing the internal bytes, not the string decimal
+        // representation. This is why we deserialize them as strings below.
+        #[derive(Deserialize)]
+        struct _Sender {
+            id: Address,
+        }
+        #[derive(Deserialize)]
+        struct _EscrowAccount {
+            balance: String,
+            #[serde(rename = "totalAmountThawing")]
+            total_amount_thawing: String,
+            sender: _Sender,
+        }
+
+        let query = GraphQLQuery {
+            query: r#"
+            query ($indexer: ID!) {
+                escrowAccounts(where: {receiver_: {id: $indexer}}) {
```

Comment:
That's because the query is doing "entity filtering", and The Graph's GraphQL API adds a `_` suffix for this: https://thegraph.com/docs/en/querying/graphql-api/#example-for-nested-entity-filtering

Reply:
PS: I was surprised at this too